Update resources.cc

ledlaux opened the pull request.

poetaster commented:
> ***@***.*** arduinoMI % git clone ***@***.***:poetaster/STMLIB.git Cloning into 'STMLIB'... ***@***.***: Permission denied (publickey). fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists. I need to generate an ssh key?
> […](#)
> On Wed, Dec 17, 2025 at 12:56 PM Vadims Maksimovs ***@***.***> wrote: Hi. I renewed plaits engine files and had to remove plaits_all.cpp which caused redefinition errors. On Wed, Dec 17, 2025 at 10:19 AM Mark Washeim ***@***.***> wrote: > *poetaster* left a comment ([poetaster/arduinoMI#12](https://github.com/poetaster/arduinoMI/pull/12)) > <[#12 (comment)](https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664183215)> > > Ok, I've tested those updates and there is something missing or I've done > an incomplete merge. These don't compile. > > What I did: > > Pull the most recent version of the Plaits1.2 branch. > Make plaits1.2 branches in PLAITS and STMLIB repos > Copy the sources from PLAITS and STMLIB from the master Plaits1.2 branch > push to the P and S plaits1.2 repos > git checkout the plaits1.2 branchs in each of STMLIB and PLAITS > > try to compile that generated a bunch of errors. > > For you to test this, go to your library folder, move PLAITS and STMLIB > out of the way and do: > > git clone ***@***.***:poetaster/STMLIB.git > git clone ***@***.***:poetaster/PLAITS.git > > cd into STMLIB > git checkout plaits1.2 > > and the same for PLAITS > > — > Reply to this email directly, view it on GitHub > <[#12 (comment)](https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664183215)>, > or unsubscribe > <https://github.com/notifications/unsubscribe-auth/AOMYGJ5ZO6P774IEM64XZ2L4CEGXXAVCNFSM6AAAAACPHANYVWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRUGE4DGMRRGU> > . > You are receiving this because you authored the thread.Message ID: > ***@***.***> >

Sorry, you can just use https: git clone https://github.com/poetaster/STMLIB.git ... etc ... bad habit

When I try to compile without the plaits_all.cpp I get:

```
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino_build_956792/sketch/PlaitsEngines.ino.cpp.o: in function `_ZN6plaits5VoiceC2Ev':
/home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x8c): undefined reference to `_ZTVN6plaits19VirtualAnalogEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x90): undefined reference to `_ZTVN6plaits17WaveshapingEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x94): undefined reference to `_ZTVN6plaits8FMEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x98): undefined reference to `_ZTVN6plaits11GrainEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x9c): undefined reference to `_ZTVN6plaits14AdditiveEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa0): undefined reference to `_ZTVN6plaits15WavetableEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa4): undefined reference to `_ZTVN6plaits11ChordEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa8): undefined reference to `_ZTVN6plaits12SpeechEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xac): undefined reference to `_ZTVN6plaits11SwarmEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb0): undefined reference to `_ZTVN6plaits11NoiseEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb4): undefined reference to `_ZTVN6plaits14ParticleEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb8): undefined reference to `_ZTVN6plaits12StringEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xbc): undefined reference to `_ZTVN6plaits11ModalEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc0): undefined reference to `_ZTVN6plaits14BassDrumEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc4): undefined reference to `_ZTVN6plaits15SnareDrumEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc8): undefined reference to `_ZTVN6plaits11HiHatEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xcc): undefined reference to `_ZTVN6plaits22VirtualAnalogVCFEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd0): undefined reference to `_ZTVN6plaits21PhaseDistortionEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd4): undefined reference to `_ZTVN6plaits11SixOpEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd8): undefined reference to `_ZTVN6plaits17WaveTerrainEngineE'
/home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino_build_956792/sketch/PlaitsEngines.ino.cpp.o: in function `_Z10initVoicesv':
```

ledlaux commented:
Yes I too get those errors if i do git pull the 1.2 platis and stmlib. I am
a bit confused. Maybe now files are mixed?

On Wed, Dec 17, 2025 at 1:20 PM Mark Washeim ***@***.***>
wrote:

> *poetaster* left a comment (poetaster/arduinoMI#12)
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664892520>
>
> *@*.*** arduinoMI % git clone *@*.*:poetaster/STMLIB.git Cloning into
> 'STMLIB'... @.*: Permission denied (publickey). fatal: Could not read
> from remote repository. Please make sure you have the correct access rights
> and the repository exists. I need to generate an ssh key?
> … <#m_-2074337303704746903_>
> On Wed, Dec 17, 2025 at 12:56 PM Vadims Maksimovs *@*.*> wrote: Hi. I
> renewed plaits engine files and had to remove plaits_all.cpp which caused
> redefinition errors. On Wed, Dec 17, 2025 at 10:19 AM Mark Washeim @.*>
> wrote: > *poetaster* left a comment (poetaster/arduinoMI#12
> <https://github.com/poetaster/arduinoMI/pull/12>) > <#12 (comment)
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664183215>>
> > > Ok, I've tested those updates and there is something missing or I've
> done > an incomplete merge. These don't compile. > > What I did: > > Pull
> the most recent version of the Plaits1.2 branch. > Make plaits1.2 branches
> in PLAITS and STMLIB repos > Copy the sources from PLAITS and STMLIB from
> the master Plaits1.2 branch > push to the P and S plaits1.2 repos > git
> checkout the plaits1.2 branchs in each of STMLIB and PLAITS > > try to
> compile that generated a bunch of errors. > > For you to test this, go to
> your library folder, move PLAITS and STMLIB > out of the way and do: > >
> git clone *@*.*:poetaster/STMLIB.git > git clone @.*:poetaster/PLAITS.git
> > > cd into STMLIB > git checkout plaits1.2 > > and the same for PLAITS > >
> — > Reply to this email directly, view it on GitHub > <#12 (comment)
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664183215>>,
> > or unsubscribe >
> https://github.com/notifications/unsubscribe-auth/AOMYGJ5ZO6P774IEM64XZ2L4CEGXXAVCNFSM6AAAAACPHANYVWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRUGE4DGMRRGU
> > . > You are receiving this because you authored the thread.Message ID: >
> *@*.***> >
>
> Sorry, you can just use https: git clone
> https://github.com/poetaster/STMLIB.git ... etc ... bad habit
>
> When I try to compile without the plaits_all.cpp I get:
>
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino_build_956792/sketch/PlaitsEngines.ino.cpp.o: in function `_ZN6plaits5VoiceC2Ev':
> /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x8c): undefined reference to `_ZTVN6plaits19VirtualAnalogEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x90): undefined reference to `_ZTVN6plaits17WaveshapingEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x94): undefined reference to `_ZTVN6plaits8FMEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x98): undefined reference to `_ZTVN6plaits11GrainEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0x9c): undefined reference to `_ZTVN6plaits14AdditiveEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa0): undefined reference to `_ZTVN6plaits15WavetableEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa4): undefined reference to `_ZTVN6plaits11ChordEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xa8): undefined reference to `_ZTVN6plaits12SpeechEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xac): undefined reference to `_ZTVN6plaits11SwarmEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb0): undefined reference to `_ZTVN6plaits11NoiseEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb4): undefined reference to `_ZTVN6plaits14ParticleEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xb8): undefined reference to `_ZTVN6plaits12StringEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xbc): undefined reference to `_ZTVN6plaits11ModalEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc0): undefined reference to `_ZTVN6plaits14BassDrumEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc4): undefined reference to `_ZTVN6plaits15SnareDrumEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xc8): undefined reference to `_ZTVN6plaits11HiHatEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xcc): undefined reference to `_ZTVN6plaits22VirtualAnalogVCFEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd0): undefined reference to `_ZTVN6plaits21PhaseDistortionEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd4): undefined reference to `_ZTVN6plaits11SixOpEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /home/mwa/Arduino/libraries/PLAITS/src/plaits/dsp/voice.h:161:(.text._ZN6plaits5VoiceC2Ev[_ZN6plaits5VoiceC5Ev]+0xd8): undefined reference to `_ZTVN6plaits17WaveTerrainEngineE'
> /home/mwa/.arduino15/packages/rp2040/tools/pqt-gcc/4.1.0-1aec55e/bin/../lib/gcc/arm-none-eabi/14.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino_build_956792/sketch/PlaitsEngines.ino.cpp.o: in function `_Z10initVoicesv':
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3664892520>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOMYGJ27EUQKF4M5JYUMBOT4CE37LAVCNFSM6AAAAACPHANYVWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRUHA4TENJSGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

poetaster commented:
> Yes I too get those errors if i do git pull the 1.2 platis and stmlib. I am a bit confused. Maybe now files are mixed? 

Not sure. I just cloned your repo, switched branch and made sure it was up to date. Maybe just package the two directories you are compiling in a tar ball and upload that here. Then I can try those 'clean'.

poetaster commented:
I did try completely replace the files with those from the zip file but that led to the errors above.

EDIT: I merged those sources into the STMLIB and PLAITS plaits1.2 branches.

ledlaux commented:
It seems that it compiles on mac by mistake not by design :D I need to look
more closely at this and start to use plaits_all.cpp imports correctly.❌
What is wrong with this PLAITS.h

You are *including implementation-level FM headers in the public API*:

#include "plaits/dsp/voice.h"#include "plaits/dsp/fm/voice.h"

This *must never happen* in Plaits 1.2.
Result

You force *both* types into the same translation unit:

   -

   plaits::Voice
   -

   plaits::fm::Voice

➡️ This *guarantees*:

reference to 'Voice' is ambiguous

Unity build or not, this will break.
------------------------------
✅ What PLAITS.h is allowed to include

PLAITS.h must be a *thin interface header only*.
✅ KEEP

#include "plaits/resources.h"#include
"plaits/dsp/engine/engine.h"#include
"plaits/dsp/engine2/six_op_engine.h"#include
"plaits/dsp/engine2/phase_distortion_engine.h"#include
"plaits/dsp/engine2/wave_terrain_engine.h"#include
"plaits/dsp/engine2/virtual_analog_vcf_engine.h"#include
"plaits/user_data.h"

❌ REMOVE (all of these)

plaits/dsp/voice.h
plaits/dsp/fm/voice.h
plaits/dsp/fm/*.h
plaits/dsp/oscillator/*.h
plaits/dsp/noise/*.h
plaits/dsp/fx/*.h
plaits/dsp/physical_modelling/*.h
plaits/dsp/drums/*.h

These belong *only* in plaits_all.cpp.
------------------------------
✅ Correct architecture (this is the key) 1️⃣ PLAITS.h

*Public API only*

   -

   engine class declarations
   -

   no DSP internals
   -

   no FM helpers
   -

   no Voice types

2️⃣ plaits_all.cpp

*Unity build*

   -

   includes *all* .cc
   -

   controls include order
   -

   defines all symbols once

------------------------------
Why this broke only now

Earlier you:

   -

   removed plaits_all.cpp
   -

   added DSP headers to PLAITS.h

That flattened the internal namespace separation Plaits relies on.

Mutable Instruments code *assumes unity build*.
------------------------------
Minimal actionable fix (do this now)

   1.

   *Delete* everything in PLAITS.h except:
   -

      resources.h
      -

      engine.h
      -

      selected engine headers
      -

      user_data.h
      2.

   Move *all DSP + FM includes* to plaits_all.cpp (as .cc)
   3.

   Ensure:

   #include "plaits/dsp/fm/voice.cc"

   comes *before* six_op_engine.cc

------------------------------
Final note (important)

This is *not* an Arduino quirk.
This is how MI Plaits is designed.

Once you do this, *all*:

   -

   Voice ambiguity
   -

   missing symbols
   -

   mac/linux differences

will disappear.

On Wed, Dec 17, 2025 at 2:23 PM Vadims Maksimovs ***@***.***> wrote:

> So I too assumed that it is definitely cc file reference and import errors
> depending on the system. For mac it works out of the box for other systems
> it should be done other way with correct cc file reference.
>
>
> On Wednesday, December 17, 2025, Vadims Maksimovs ***@***.***>
> wrote:
>
>> This clue from chatgpt maybe can help:
>>
>>
>> Short version (what’s actually wrong):
>>
>> You mixed two different Plaits code models.
>> That’s why you’re seeing:
>>
>> Voice is ambiguous
>>
>> missing .cc symbols
>>
>> need for / removal of plaits_all.cpp depending on machine
>>
>>
>> Root cause
>>
>> Plaits 1.2 expects unity build style:
>>
>> plaits_all.cpp must exist
>>
>> engine .cc files are not compiled individually
>>
>> FM engines use both:
>>
>> plaits::Voice
>>
>> plaits::fm::Voice
>>
>>
>>
>> When you removed plaits_all.cpp and started including .cc files manually,
>> Arduino started compiling things twice and lost namespace context →
>> ambiguity errors.
>>
>> macOS “working” was just luck in compile order.
>>
>>
>> ---
>>
>> Correct fix (do this exactly)
>>
>> 1. Restore plaits_all.cpp
>>
>> From original Plaits 1.2 repo. Do not modify it.
>>
>> 2. Remove all manual .cc includes
>>
>> In:
>>
>> PLAITS.h
>>
>> resources.cc
>>
>> anywhere else
>>
>>
>> ❌ no #include "six_op_engine.cc"
>> ❌ no engine .cc in Arduino sketch
>>
>> 3. Arduino library layout must be:
>>
>> Arduino/libraries/
>>  ├─ STMLIB/
>>  │ └─ src/...
>>  └─ PLAITS/
>>      └─ src/
>>          ├─ plaits_all.cpp
>>          ├─ plaits/
>>          └─ resources.cc
>>
>> Only .h files are included by users.
>>
>> 4. Do NOT pull upstream Plaits partially
>>
>> Either:
>>
>> use pure Plaits 1.2 snapshot or
>>
>> fully merge all engine + fm + dsp changes
>>
>>
>> Partial pulls = guaranteed breakage.
>>
>>
>> ---
>>
>> Why the Voice error happens
>>
>> Voice<6>::Parameters
>>
>> is ambiguous because:
>>
>> plaits::Voice
>>
>> plaits::fm::Voice
>>
>>
>> Unity build (plaits_all.cpp) sets correct include order and namespace
>> visibility. Without it → compiler confusion.
>>
>>
>>
>>
>> On Wednesday, December 17, 2025, Mark Washeim ***@***.***>
>> wrote:
>>
>>> *poetaster* left a comment (poetaster/arduinoMI#12)
>>> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3665059780>
>>>
>>> I did try completely replace the files with those from the zip file but
>>> that led to the errors above.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3665059780>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AOMYGJ4GD37JJBNYFY7VAVT4CFB27AVCNFSM6AAAAACPHANYVWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRVGA2TSNZYGA>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> ***@***.***>
>>>
>>

poetaster commented:
I will also be removing user data everywhere. That's a construct that only makes sense in the original hardware since it's for saving presets on given hardware.

poetaster commented:
For reference, the ugens for super collider are built with cmake. That's where I got the original includes from. They work on all operating system:

```set(STMLIB_SOURCES
    ${STMLIB_PATH}/stmlib.h
    ${STMLIB_PATH}/utils/random.cc
    ${STMLIB_PATH}/utils/random.h
    ${STMLIB_PATH}/utils/dsp.h
    ${STMLIB_PATH}/dsp/atan.cc
    ${STMLIB_PATH}/dsp/atan.h
    ${STMLIB_PATH}/dsp/units.cc
    ${STMLIB_PATH}/dsp/units.h

)
set(MI_SOURCES
    ${MI_PATH}/resources.cc
    ${MI_PATH}/resources.h
    ${MI_PATH}/dsp/voice.cc
    ${MI_PATH}/dsp/voice.h
    ${MI_PATH}/dsp/speech/lpc_speech_synth.cc
    ${MI_PATH}/dsp/speech/lpc_speech_synth.h
    ${MI_PATH}/dsp/speech/lpc_speech_synth_controller.cc
    ${MI_PATH}/dsp/speech/lpc_speech_synth_controller.h
    ${MI_PATH}/dsp/speech/lpc_speech_synth_phonemes.cc
    ${MI_PATH}/dsp/speech/lpc_speech_synth_words.cc
    ${MI_PATH}/dsp/speech/lpc_speech_synth_words.h
    ${MI_PATH}/dsp/speech/naive_speech_synth.cc
    ${MI_PATH}/dsp/speech/naive_speech_synth.h
    ${MI_PATH}/dsp/speech/sam_speech_synth.cc
    ${MI_PATH}/dsp/speech/sam_speech_synth.h
    ${MI_PATH}/dsp/drums/analog_bass_drum.h
    ${MI_PATH}/dsp/drums/analog_snare_drum.h
    ${MI_PATH}/dsp/drums/hi_hat.h
    ${MI_PATH}/dsp/drums/synthetic_bass_drum.h
    ${MI_PATH}/dsp/drums/synthetic_snare_drum.h
    ${MI_PATH}/dsp/dsp.h
    ${MI_PATH}/dsp/engine/additive_engine.cc
    ${MI_PATH}/dsp/engine/additive_engine.h
    ${MI_PATH}/dsp/engine/bass_drum_engine.cc
    ${MI_PATH}/dsp/engine/bass_drum_engine.h
    ${MI_PATH}/dsp/engine/chord_engine.cc
    ${MI_PATH}/dsp/engine/chord_engine.h
    ${MI_PATH}/dsp/engine/engine.h
    ${MI_PATH}/dsp/engine/fm_engine.cc
    ${MI_PATH}/dsp/engine/fm_engine.h
    ${MI_PATH}/dsp/engine/grain_engine.cc
    ${MI_PATH}/dsp/engine/grain_engine.h
    ${MI_PATH}/dsp/engine/hi_hat_engine.cc
    ${MI_PATH}/dsp/engine/hi_hat_engine.h
    ${MI_PATH}/dsp/engine/modal_engine.cc
    ${MI_PATH}/dsp/engine/modal_engine.h
    ${MI_PATH}/dsp/engine/noise_engine.cc
    ${MI_PATH}/dsp/engine/noise_engine.h
    ${MI_PATH}/dsp/engine/particle_engine.cc
    ${MI_PATH}/dsp/engine/particle_engine.h
    ${MI_PATH}/dsp/engine/snare_drum_engine.cc
    ${MI_PATH}/dsp/engine/snare_drum_engine.h
    ${MI_PATH}/dsp/engine/speech_engine.cc
    ${MI_PATH}/dsp/engine/speech_engine.h
    ${MI_PATH}/dsp/engine/string_engine.cc
    ${MI_PATH}/dsp/engine/string_engine.h
    ${MI_PATH}/dsp/engine/swarm_engine.cc
    ${MI_PATH}/dsp/engine/swarm_engine.h
    ${MI_PATH}/dsp/engine/virtual_analog_engine.cc
    ${MI_PATH}/dsp/engine/virtual_analog_engine.h
    ${MI_PATH}/dsp/engine/waveshaping_engine.cc
    ${MI_PATH}/dsp/engine/waveshaping_engine.h
    ${MI_PATH}/dsp/engine/wavetable_engine.cc
    ${MI_PATH}/dsp/engine/wavetable_engine.h
    ${MI_PATH}/dsp/envelope.h
  ${MI_PATH}/dsp/fx/diffuser.h
    ${MI_PATH}/dsp/fx/fx_engine.h
    ${MI_PATH}/dsp/fx/low_pass_gate.h
    ${MI_PATH}/dsp/fx/overdrive.h
    ${MI_PATH}/dsp/fx/sample_rate_reducer.h
    ${MI_PATH}/dsp/noise/clocked_noise.h
    ${MI_PATH}/dsp/noise/dust.h
    ${MI_PATH}/dsp/noise/fractal_random_generator.h
    ${MI_PATH}/dsp/noise/particle.h
    ${MI_PATH}/dsp/noise/smooth_random_generator.h
    ${MI_PATH}/dsp/oscillator/formant_oscillator.h
    ${MI_PATH}/dsp/oscillator/grainlet_oscillator.h
    ${MI_PATH}/dsp/oscillator/harmonic_oscillator.h
    ${MI_PATH}/dsp/oscillator/oscillator.h
    ${MI_PATH}/dsp/oscillator/sine_oscillator.h
    ${MI_PATH}/dsp/oscillator/string_synth_oscillator.h
    ${MI_PATH}/dsp/oscillator/variable_saw_oscillator.h
    ${MI_PATH}/dsp/oscillator/variable_shape_oscillator.h
    ${MI_PATH}/dsp/oscillator/vosim_oscillator.h
    ${MI_PATH}/dsp/oscillator/wavetable_oscillator.h
    ${MI_PATH}/dsp/oscillator/z_oscillator.h
    ${MI_PATH}/dsp/physical_modelling/delay_line.h
    ${MI_PATH}/dsp/physical_modelling/modal_voice.cc
    ${MI_PATH}/dsp/physical_modelling/modal_voice.h
    ${MI_PATH}/dsp/physical_modelling/resonator.cc
    ${MI_PATH}/dsp/physical_modelling/resonator.h
    ${MI_PATH}/dsp/physical_modelling/string.cc
    ${MI_PATH}/dsp/physical_modelling/string.h
    ${MI_PATH}/dsp/physical_modelling/string_voice.cc
    ${MI_PATH}/dsp/physical_modelling/string_voice.h
)
```
That looks even messier ;)

ledlaux commented:
> Ok, I can't follow your instructions. Could you please clone the PLAITS and STMLIB modules and implement them in the plaits1.2 branches there?

From which branch to clone them?

ledlaux commented:
So i did this:
1) Cloned arduinoMI repo.
2) Copied smtlib and plaits folders to arduino/libs
3) removed all .inc files, smtlib_all.cpp, plaits_all.cpp
4) replaced all files in folder PLAITS with files from folder PLAITS from my plaits1.2 zip file
5) PlaitsEngines example compailed without error. 

if I leave smtlib_all.cpp and plaits_all.cpp files and .inc files there I got bunch of redefinition errors. For now I cant make it work as intended with those files in place and .inc file system.

poetaster commented:
With the checkout mechanism I can simply change which version works in which context. 
Plaits 'main' branch has only 2 extra engines (I backported) but compiles for simple cases (all ide versions)
Plaits 'plaits1.0mmm' has 3 extra backported engines and compiles with MMM in all ide versions.

ledlaux commented:
Just wanted to know if it works for others too :)

On Wed, Dec 17, 2025 at 4:42 PM Mark Washeim ***@***.***>
wrote:

> *poetaster* left a comment (poetaster/arduinoMI#12)
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3665671653>
>
> It's all good! The whole point of the project, for me, was to make MI more
> usable. You got there, so all's good :) There are a number of other people
> (one on a mac) using some or all of the libraries so I'm happy :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/poetaster/arduinoMI/pull/12#issuecomment-3665671653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOMYGJYZZTLYGAPW6B2CDNL4CFTWLAVCNFSM6AAAAACPHANYVWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMNRVGY3TCNRVGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

poetaster commented:
I just pushed an update to the PLAITS repo, plaits1.2 branch which compiles. However, it crashes immediately :) But at least it compiles :)

poetaster commented:
I did have another idea. Make a new module called SixOp and dedicate it to FM only. Then one could have a huge number of DEX llike patches. I have DEX running on old raspberry pi 1 computers, so I don't really need one, but it would be nice to have a dedicated FM machine in eurorack format......

ledlaux commented:
To add more dx7 banks to plaits we need to:
1. Add syx bank array to resources.cc
2. Add in resources.cc
  const uint8_t* const fm_patches_table[] = {
  syx_bank_0,
  syx_bank_1,
  syx_bank_2,
  syx_bank_3,
};

3. Add extern const uint8_t syx_bank_3[]; in resources.h
4. Add case 3: return syx_bank_3; in user_data.h
5. Add engines_.RegisterInstance(&six_op_engine_3, true, 1.0f, 1.0f); in voice.cc
6.  Add SixOpEngine six_op_engine_3; in voice.h

Must be it for adding new banks, but I have not tested how many we can add more.
